homeassistant/components/nextcloud/sensor.py

```diff
@@ -0,0 +1,111 @@
+"""Summary data from Nextcoud."""
+from datetime import timedelta
+import logging
+
+from nextcloudmonitor import NextcloudMonitor
+import voluptuous as vol
+
+from homeassistant.components.sensor import PLATFORM_SCHEMA
+from homeassistant.const import (
+    CONF_PASSWORD,
+    CONF_SCAN_INTERVAL,
+    CONF_URL,
+    CONF_USERNAME,
+)
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.event import track_time_interval
+
+_LOGGER = logging.getLogger(__name__)
+
+# Assign domain
+DOMAIN = "nextcloud"
+# Set Default scan interval in seconds
+SCAN_INTERVAL = timedelta(seconds=60)
+
+# Validate user configuration
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
+    {
+        vol.Required(CONF_URL): cv.url,
+        vol.Required(CONF_USERNAME): cv.string,
+        vol.Required(CONF_PASSWORD): cv.string,
+        vol.Optional(CONF_SCAN_INTERVAL, default=SCAN_INTERVAL): cv.time_period,
+    }
+)
+
+
+def setup_platform(hass, config, add_entities, discovery_info=None) -> None:
+    """Set up the Nextcloud sensor."""
+    # Fetch Nextcloud Monitor api data
+    ncm = NextcloudMonitor(
+        config[CONF_URL], config[CONF_USERNAME], config[CONF_PASSWORD]
+    )
+    hass.data[DOMAIN] = get_sensors(ncm.data)
+
+    def nextcloud_update(event_time):
+        """Update data from nextcloud api."""
+        ncm.update()
+        hass.data[DOMAIN] = get_sensors(ncm.data)
+
+    # Update sensors on time interval
+    track_time_interval(hass, nextcloud_update, config[CONF_SCAN_INTERVAL])
```

Comment:
We prefer to use our data coordinator helper if all data can be fetched in a single api call.

https://developers.home-assistant.io/docs/integration_fetching_data#coordinated-single-api-poll-for-data-for-all-entities

Reply:
Yes, all data can be fetched for all sensors in  a single call. 
Basically the api call returns a json object containing all sensor data. The nextcloud component here parses the response and returns sensors based on the json endpoints (leaves). The names of the sensors are the path of the json endpoint and the state of the sensor is the value of the endpoint itself.

Reply:
@MartinHjelmare Does the data coordinator approach require async?
I tried implementing it without async, but it looks like it only has an `async_refresh` method.

Reply:
Yes it need only has an async interface. It's usually easier then to make the top level context of your integration async and use and await `hass.async_add_executor_job` to schedule sync work to the thread pool as needed.